### PR TITLE
Add Azure message converter

### PR DESCRIPTION
**Pull request by warrenzhu25**

### CODE REVIEW

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+
+/**
+ * Abstract class handles common conversion logic between <T> and {@link Message}
+ *
+ * @author Warren Zhu
+ */
+public abstract class AbstractAzureMessageConverter<T> implements AzureMessageConverter<T> {
+    private static ObjectMapper objectMapper = new ObjectMapper();
```
Comment (sophiaso) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: Whether we need to consider allow user to customize this `objectMapper`, in order to support different data mapping requirements?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+        return (Message<U>) internalToMessage(azureMessage, headers, targetPayloadClass);
+    }
+
+    protected abstract byte[] getPayload(T azureMessage);
+
+    protected abstract T fromString(String payload);
+
+    protected abstract T fromByte(byte[] payload);
+
+    protected void setCustomHeaders(Message<?> message, T azureMessage) {
```
Comment (sophiaso) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: This method is empty, add `TODO`?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+    private <U> Message<?> internalToMessage(T azureMessage, MessageHeaders headers, Class<U> targetPayloadClass) {
+        byte[] payload = getPayload(azureMessage);
+
+        if (targetPayloadClass.isInstance(azureMessage)) {
+            return MessageBuilder.withPayload(azureMessage).copyHeaders(headers).build();
+        }
+
+        if (targetPayloadClass == String.class) {
+            return MessageBuilder.withPayload(new String(payload)).copyHeaders(headers).build();
```
Comment (sophiaso) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: Why the compare here is `exactly equals` String.class (same below), but internalFromMessage is checking with `instance of`?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+/**
+ * Abstract class handles common conversion logic between <T> and {@link Message}
+ *
+ * @author Warren Zhu
+ */
+public abstract class AbstractAzureMessageConverter<T> implements AzureMessageConverter<T> {
+    private static ObjectMapper objectMapper = new ObjectMapper();
+
+    @Override
+    public T fromMessage(Message<?> message, Class<T> targetClass) {
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: Would message conversion failure be possible? Shall we have a MessageConversionException anyway?

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 2 above)
Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: I think it is a hook method users can override.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <U> Message<U> toMessage(T azureMessage, MessageHeaders headers, Class<U> targetPayloadClass) {
+        return (Message<U>) internalToMessage(azureMessage, headers, targetPayloadClass);
+    }
+
+    protected abstract byte[] getPayload(T azureMessage);
+
+    protected abstract T fromString(String payload);
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: how do we handle payload that's massive? we should treat the payload in the way of stream regardless it's text message or byte message.

e.g. MarshallingMessageConverter

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 1 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: We could provide a setter to let users provide their `objectMapper`.  But in this version, I want to keep it simply.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 2 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: Yes.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 3 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: This comparison is between both Class literal. Another is between object and Class literal.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 4 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: We have `ConversionException` here.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java (hunk 5 above)
Comment (warrenzhu25) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: 1. If payload is too large, then backed service should throw exception. Normally, service has its constraints of message size. 
2. `Stream` also has text-based and byte-based, we still need to distinguish.

Comment (yaweiw) [reply spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/AbstractAzureMessageConverter.java]: my ask is: we should we look into org.springframework.jms.support.converter. See if we could do better. Message conversion will be widely used in our repo.

File: spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/ConversionException.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ * @author Warren Zhu
+ */
+public class ConversionException extends NestedRuntimeException {
+
+    public ConversionException(String msg) {
+        super(msg);
+    }
+
+    public ConversionException(String msg, Throwable cause) {
```
Comment (yaweiw) [inline spring-integration-azure/src/main/java/com/microsoft/azure/spring/integration/core/converter/ConversionException.java]: cause @Nullable?

**Pull request merged**